### PR TITLE
timesync: Make delaying attempts to contact servers configurable

### PR DESCRIPTION
**Pull request by ssahani**
```
❯ ssh sus@xx.xx.xx.xx
Last login: Sat Nov 14 17:32:08 2020 from 10.104.45.138
 17:36:19 up 0 min,  0 users,  load average: 0.00, 0.00, 0.00
> systemd-analyze blame
Bootup is not yet finished (org.freedesktop.systemd1.Manager.FinishTimestampMonotonic=0).
Please try again later.
Hint: Use 'systemctl list-jobs' to see active jobs
> systemd-analyze blame
43.954s systemd-time-wait-sync.service
 1.969s systemd-networkd-wait-online.service
 1.559s cloud-init-local.service
 1.039s cloud-init.service
  414ms cloud-final.service
  387ms dracut-initqueue.service
  382ms initrd-switch-root.service
  380ms cloud-config.service
  198ms systemd-journal-flush.service
  136ms systemd-udev-trigger.service
  115ms initrd-parse-etc.service
   97ms systemd-timesyncd.service
   84ms systemd-journald.service
```

When set to 5s

```
❯ ssh sus@xx.xx.xx.xx
Last login: Sat Nov 14 18:41:42 2020 from 10.104.45.138
 18:42:36 up 0 min,  0 users,  load average: 0.16, 0.03, 0.01
> systemd-analyze blame
10.450s systemd-time-wait-sync.service
 8.303s systemd-networkd-wait-online.service
 1.621s cloud-init-local.service
 1.068s cloud-init.service
```

### CODE REVIEW

File: src/timesync/timesyncd-manager.c [hunk 1]
```diff
@@ -787,7 +786,8 @@ int manager_connect(Manager *m) {
         if (!ratelimit_below(&m->ratelimit)) {
                 log_debug("Delaying attempts to contact servers.");
 
-                r = sd_event_add_time_relative(m->event, &m->event_retry, clock_boottime_or_monotonic(), RETRY_USEC, 0, manager_retry_connect, m);
+                r = sd_event_add_time_relative(m->event, &m->event_retry, clock_boottime_or_monotonic(), m->network_online_retry_timer,
```
Comment (bluca) [inline src/timesync/timesyncd-manager.c]: In my testing and reproing of the issue, this code path is not hit at all. The default 30s would not be a problem here - the problem is that we are hitting PollIntervalMaxSec - which is half an hour - and never backing down from there, because networkd is not used and thus we can't detect that the network is gone/back and react.

Comment (bluca) [reply src/timesync/timesyncd-manager.c]: In other words, in steady state we get poll_interval_usec all the way up to PollIntervalMaxSec, then the network goes away, and we have nothing to wake us up before it expires.

That is at least my understanding, which might very well be wrong.

Comment (bluca) [reply src/timesync/timesyncd-manager.c]: Basing this on observing what happens and debug level logs:

```
Dec 08 14:22:32 debian systemd-timesyncd[3679]: Selected server 0.debian.pool.ntp.org.
Dec 08 14:22:32 debian systemd-timesyncd[3679]: Resolving 0.debian.pool.ntp.org...
Dec 08 14:22:38 debian systemd-timesyncd[3679]: Failed to resolve 0.debian.pool.ntp.org: Name or service not known
Dec 08 14:22:38 debian systemd-timesyncd[3679]: Selected server 1.debian.pool.ntp.org.
Dec 08 14:22:38 debian systemd-timesyncd[3679]: Resolving 1.debian.pool.ntp.org...
Dec 08 14:22:44 debian systemd-timesyncd[3679]: Failed to resolve 1.debian.pool.ntp.org: Name or service not known
Dec 08 14:22:44 debian systemd-timesyncd[3679]: Selected server 2.debian.pool.ntp.org.
Dec 08 14:22:44 debian systemd-timesyncd[3679]: Resolving 2.debian.pool.ntp.org...
Dec 08 14:22:50 debian systemd-timesyncd[3679]: Failed to resolve 2.debian.pool.ntp.org: Name or service not known
Dec 08 14:22:50 debian systemd-timesyncd[3679]: Selected server 3.debian.pool.ntp.org.
Dec 08 14:22:50 debian systemd-timesyncd[3679]: Resolving 3.debian.pool.ntp.org...
Dec 08 14:22:56 debian systemd-timesyncd[3679]: Failed to resolve 3.debian.pool.ntp.org: Name or service not known
Dec 08 14:22:56 debian systemd-timesyncd[3679]: Waiting after exhausting servers.
```

Comment (ssahani) [reply src/timesync/timesyncd-manager.c]: Your testing is wrong. Netword is enabled by default

Comment (bluca) [reply src/timesync/timesyncd-manager.c]: It is not on Debian, which is where https://github.com/systemd/systemd/issues/16529 was reported. On Debian, ifupdown is used on headless installations and network-manager on desktops.

Comment (ssahani) [reply src/timesync/timesyncd-manager.c]: You should test on system where netword is enabled else its pointless

File: man/timesyncd.conf.xml [hunk 2]
```diff
@@ -91,6 +91,12 @@
         <varname>PollIntervalMaxSec=</varname> defaults to 2048 seconds.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>ContactServersRetryIntervalSec</varname></term>
```
Comment (poettering) [inline man/timesyncd.conf.xml]: Maybe call this ConnectionRetrySec=?

File: man/timesyncd.conf.xml [hunk 3]
```diff
@@ -91,6 +91,12 @@
...
+        <listitem><para>Specifies the delaying attempts to contact servers after network is online. Takes a time value (in seconds).
+        Defaults to 30 seconds and must not be smaller than 5 seconds.</para></listitem>
```
Comment (poettering) [inline man/timesyncd.conf.xml]: why the 5s limit? that sounds unnecessary? 1s should work too?

**Review by poettering — CHANGES_REQUESTED** [review]
looks ok, but I'd prefer a short name fot he option, see above

**Comment by poettering** [comment]
so, independently of whether this actually fixes #16529, I think making this parameter configurable probably makes sense. hence i am sympathetic to the commit, but please rename the option.

That said, to me too it appears the fix for #16529 is someting else, the hostname resolution failure appears not to be handled properly

File: man/timesyncd.conf.xml (hunk 3 above)
Comment (ssahani) [reply man/timesyncd.conf.xml]: I tested it 10/15 times. Something below 5 it's gets into a loop sometimes and gets too bz. system does not boot and hangs.

**Comment by bluca** [comment]
> so, independently of whether this actually fixes #16529, I think making this parameter configurable probably makes sense. hence i am sympathetic to the commit, but please rename the option.
> 
> That said, to me too it appears the fix for #16529 is someting else, the hostname resolution failure appears not to be handled properly

Sounds good to me for this PR, if the setting is useful by itself I've got no objections.

For #16529 when I reproduced it I had the distinct impression that the issue is not so much about shortening timers, but getting kicked into action. As far as I can see, when the interval goes all the way up to the max because everything is good, and networkd is not used so there's no notification on interface state change, there's nothing that can wake timesyncd and get it to look at things again. Once it does and it sees that the servers are unreachable then shortening the timer will help making things happen quicker once they are fixed, true, but still it will take time (and add pointless load on the system in the meanwhile).
Wouldn't it be better to add ExecReload support or something like that, which would kick timesyncd back into action, and document it?

Then in distros like Debian, where we ship with timesyncd but not with networkd by default, we can add ifupdown/network-manager/whatever hooks. And everything should start working immediately on state change, just like it does when using networkd.

**Comment by ssahani** [comment]
Updated thanks for the review.

File: man/timesyncd.conf.xml [hunk 4]
```diff
@@ -91,6 +91,12 @@
         <varname>PollIntervalMaxSec=</varname> defaults to 2048 seconds.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>ConnectionRetrySec</varname></term>
```
Comment (poettering) [inline man/timesyncd.conf.xml]: please suffix with "=" like we usually do for configuration variables that expect a value. i.e.

`<varname>ConnectionRetrySec</varname>` → `ConnectionRetrySec=</varname>`

File: src/timesync/timesyncd-conf.c [hunk 5]
```diff
@@ -124,5 +124,10 @@ int manager_parse_config_file(Manager *m) {
                 m->poll_interval_max_usec = MAX(NTP_POLL_INTERVAL_MAX_USEC, m->poll_interval_min_usec * 32);
         }
 
+        if (m->network_online_retry_timer < 1 * USEC_PER_SEC) {
+                log_warning("Invalid ConnectionRetrySec= . Using default value.");
```
Comment (poettering) [inline src/timesync/timesyncd-conf.c]: the space after = and before . looks misplaced, please remove.

File: src/timesync/timesyncd-manager.h [hunk 6]
```diff
@@ -27,6 +27,8 @@ typedef struct Manager Manager;
 #define NTP_RETRY_INTERVAL_MIN_USEC     (15 * USEC_PER_SEC)
 #define NTP_RETRY_INTERVAL_MAX_USEC     (6 * 60 * USEC_PER_SEC) /* 6 minutes */
 
+#define RETRY_USEC (30*USEC_PER_SEC)
```
Comment (poettering) [inline src/timesync/timesyncd-manager.h]: please name this DEFAULT_CONNECTION_RETRY_USEC to keep the naming in sync with the user-facing config option (of course they'll have a different sec/usec suffix, but that's fine)

File: src/timesync/timesyncd-manager.c [hunk 7]
```diff
@@ -1085,6 +1085,8 @@ int manager_new(Manager **ret) {
         m->poll_interval_min_usec = NTP_POLL_INTERVAL_MIN_USEC;
         m->poll_interval_max_usec = NTP_POLL_INTERVAL_MAX_USEC;
 
+        m->network_online_retry_timer = RETRY_USEC;
```
Comment (poettering) [inline src/timesync/timesyncd-manager.c]: connection_retry_timer_usec, please

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/timesync/timesyncd-manager.h [hunk 8]
```diff
@@ -60,6 +62,7 @@ struct Manager {
         struct timespec trans_time_mon;
         struct timespec trans_time;
         usec_t retry_interval;
+        usec_t connection_retry_timer_usec;
```
Comment (poettering) [inline src/timesync/timesyncd-manager.h]: i'd drop the "_timer_" too from this.

Comment (ssahani) [reply src/timesync/timesyncd-manager.h]: Done thanks !

File: src/network/networkd-manager.c [hunk 9]
```diff
@@ -313,6 +313,9 @@ static int manager_rtnl_process_link(sd_netlink *rtnl, sd_netlink_message *messa
                         /* link is new, so add it */
                         r = link_add(m, message, &link);
                         if (r < 0) {
+                                if (r == -ENODEV)
+                                        link_drop(link);
```
Comment (bluca) [inline src/network/networkd-manager.c]: Mmh this change does not seem related to the rest? Or is it?

Comment (ssahani) [reply src/network/networkd-manager.c]: Ouch I pushed in the wrong tab.

**Comment by bluca** [comment]
Looks good now, all comments resolved. Will merge tomorrow morning if there are no more comments from @poettering

**Comment by yuwata** [comment]
LGTM too.
